pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -816,45 +816,77 @@ protected int getKafkaPartitionCount(StreamConfig streamConfig) {
-   * @return
+   * Returns the LLC realtime segments for the given table.
+   *
+   * @param realtimeTableName Realtime table name
+   * @return List of LLC realtime segment names
    */
   @VisibleForTesting
-  protected List<LLCRealtimeSegmentZKMetadata> getAllSegmentMetadata(String tableNameWithType) {
-    return ZKMetadataProvider.getLLCRealtimeSegmentZKMetadataListForTable(_helixManager.getHelixPropertyStore(),
-        tableNameWithType);
+  protected List<String> getAllSegments(String realtimeTableName) {
+    return ZKMetadataProvider.getLLCRealtimeSegments(_propertyStore, realtimeTableName);
   }
 
   /**
-   * Gets latest 2 metadata. We need only the 2 latest metadata for each partition in order to perform repairs
-    * @param tableNameWithType
-   * @return
+   * Returns the LLC realtime segment ZK metadata for the given table and segment.
+   *
+   * @param realtimeTableName Realtime table name
+   * @param segmentName Segment name (String)
+   * @return LLC realtime segment ZK metadata
    */
   @VisibleForTesting
-  protected Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> getLatestMetadata(String tableNameWithType) {
-    List<LLCRealtimeSegmentZKMetadata> metadatas = getAllSegmentMetadata(tableNameWithType);
-
-    Comparator<LLCRealtimeSegmentZKMetadata> comparator = (o1, o2) -> {
-      LLCSegmentName s1 = new LLCSegmentName(o1.getSegmentName());
-      LLCSegmentName s2 = new LLCSegmentName(o2.getSegmentName());
-      return s2.compareTo(s1);
-    };
-
-    Map<Integer, MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata>> partitionToLatestSegments = new HashMap<>();
-
-    for (LLCRealtimeSegmentZKMetadata metadata : metadatas) {
-      LLCSegmentName segmentName = new LLCSegmentName(metadata.getSegmentName());
-      final int partitionId = segmentName.getPartitionId();
-      MinMaxPriorityQueue<LLCRealtimeSegmentZKMetadata> latestSegments = partitionToLatestSegments.get(partitionId);
-      if (latestSegments == null) {
-        latestSegments = MinMaxPriorityQueue.orderedBy(comparator).maximumSize(2).create();
-        partitionToLatestSegments.put(partitionId, latestSegments);
+  protected LLCRealtimeSegmentZKMetadata getSegmentMetadata(String realtimeTableName, String segmentName) {
+    return (LLCRealtimeSegmentZKMetadata) ZKMetadataProvider.getRealtimeSegmentZKMetadata(_propertyStore,
+        realtimeTableName, segmentName);
+  }
+
+  /**
+   * Returns the latest 2 LLC realtime segment ZK metadata for each partition.
+   *
+   * @param realtimeTableName Realtime table name
+   * @return Map from partition to list of latest LLC realtime segment ZK metadata
+   */
+  @VisibleForTesting
+  protected Map<Integer, List<LLCRealtimeSegmentZKMetadata>> getLatestMetadata(String realtimeTableName) {
+    List<String> llcRealtimeSegments = getAllSegments(realtimeTableName);
+
+    Comparator<LLCSegmentName> segmentNameComparator = (o1, o2) -> o1.getSequenceNumber() - o2.getSequenceNumber();
```

Comment:
why not invert the comparator, and just do q.offer (instead of checking size etc.)?
You will have to shuffle the order at the end, but you are putting it into a list anyway, so no extra logic.

Reply:
Time complexity is very different. One is O(n) but another one is O(nlogn)